### PR TITLE
HGCAL simulation of ToA constant jitter noise term

### PR DESCRIPTION
**Pull request by stahlleiton**
#### PR description:

This PR introduces the simulation of the constant term of the jitter noise on the time of arrival (ToA) measurement of the HGCAL. This is done by modelling a Gaussian noise term common to all HGCAL hits and added as a bias to the ToA.
This is part of the set of updates already announced in the HGCAL DPG meeting (see https://indico.cern.ch/event/1158758/contributions/4880969/attachments/2451390/4201020/Timining%2020220525%20v2-2.pdf).

#### PR validation:

PR was validated using the runTheMatrix.py -w upgrade -l 23234.103 test.
HGCAL-related changes were mainly observed on the ToA distribution as expected while changes in the ADC are due to different number of calls of the random number generator, which causes some smearing visible at lower ADC values.

Plots of ToA for jitter constant of 20 ps: 
image
Plots of ToA for jitter constant of 1000 ps: 
image

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

### CODE REVIEW

**Comment by pfs** [comment]
assign hgcal-dpg

**Comment by pfs** [comment]
please test

**Comment by felicepantaleo** [comment]
@stahlleiton thanks for this PR. 
I add some trivial technical changes to the review.

Did you have the chance to look at what happens downstream? If you want, as further validation that what you are doing is correct, but also to "warn" other POGs using timing.
I see some suspicious new entries in the ECAL PFClusters timing at 10^38...
image

**Review by felicepantaleo — CHANGES_REQUESTED** [review]

**Comment by civanch** [comment]
please test

File: SimCalorimetry/HGCalSimProducers/src/HGCFEElectronics.cc [hunk 1]
```diff
@@ -248,11 +250,14 @@ void HGCFEElectronics<DFr>::runShaperWithToT(DFr& dataFrame,
...
     timeToA = toaColl[fireBX];
-    float jitter = getTimeJitter(chargeColl[fireBX], thickness);
+    float sensor_noise = noiseWidth <= 0 ? noise_fC_.at(thickness - 1) : noiseWidth;
+    float noise = std::sqrt(jitterNoise2_ns_.at(thickness - 1)) * sensor_noise;
+    float jitter = chargeColl[fireBX] == 0 ? 0 : (noise / chargeColl[fireBX]);
     if (jitter != 0)
       timeToA = CLHEP::RandGaussQ::shoot(engine, timeToA, jitter);
     else if (tdcResolutionInNs_ != 0)
       timeToA = CLHEP::RandGaussQ::shoot(engine, timeToA, tdcResolutionInNs_);
+    timeToA += eventTimeOffset_ns_.at(thickness - 1);
```
[the remark this answers is not in the document]
Comment (stahlleiton) [reply SimCalorimetry/HGCalSimProducers/src/HGCFEElectronics.cc]: changed all the .at() with []

File: SimCalorimetry/HGCalSimProducers/src/HGCFEElectronics.cc [hunk 2]
```diff
@@ -248,11 +250,14 @@ void HGCFEElectronics<DFr>::runShaperWithToT(DFr& dataFrame,
   //to be done properly with realistic ToA shaper and jitter for the moment accounted in the smearing
   if (toaColl[fireBX] != 0.f) {
     timeToA = toaColl[fireBX];
-    float jitter = getTimeJitter(chargeColl[fireBX], thickness);
+    float sensor_noise = noiseWidth <= 0 ? noise_fC_.at(thickness - 1) : noiseWidth;
+    float noise = std::sqrt(jitterNoise2_ns_.at(thickness - 1)) * sensor_noise;
```
[the remark this answers is not in the document]
Comment (stahlleiton) [reply SimCalorimetry/HGCalSimProducers/src/HGCFEElectronics.cc]: changed the jitterNoise2_ns_ and jitterConstant2_ns_ to their normal (not squared) versions jitterNoise_ns_ and jitterConstant_ns_

**Comment by civanch** [comment]
+1

**Comment by felicepantaleo** [comment]
+1

**Comment by sameasy** [comment]
Hi, The differences seen when testing this PR seem to be not restricted to ECAL information only, but also appear in other sub-systems as well. There seems to be no migration of clusters between HGCAL and ECAL/HCAL etc. In order to check if there were any issues in the unit tests, I ran a test workflow [runTheMatrix.py -w upgrade -l 23234.103] with and without the PR. Differences persist still and can be seen in rechit collections from ecal, hcal and mtd from the attached plots (red is before and blue is after the PR). It seems that the issue may be with the random number generators. They should be distinct but it seems like there is some dependency on the calls made by the sub-systems.

For the issue about the the appearance of new entries at very high ECAL time values I find that the high values are actually there even without the PR. So this may point to some other issue unrelated to the PR.
mtdrechits
hcalrechits
ecalrechits
timedist

**Comment by srimanob** [comment]
@cmsbuild please test
Just re-trigger the test after 3 weeks. Then we can sign.

**Comment by pfs** [comment]
@srimanob  Is it ok to approve now that the tests have run?

**Comment by srimanob** [comment]
+Upgrade

**Comment by perrotta** [comment]
+1

**Pull request merged**